staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// Decorator composes an inner http.ResponseWriter object, the decorator
+// object should return the inner (decorated) http.ResponseWriter
+// object via the Unwrap function.
+type Decorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated with the decorator
+	Unwrap() http.ResponseWriter
+}
+
+// Wrap accepts a Decorator object and wraps with an "outer" http.ResponseWriter
+// object that implements http.CloseNotifier, http.Flusher and http.Hijacker accordingly
+//
+// We use three terms here:
+//  - "outer": this is the ResponseWriter oject returned by this function. The outer object
+//    must implement the same exact interfaces as does the inner object. Our scope is limited
+//    to http.CloseNotifier, http.Flusher and http.Hijacker. This function does not take into
+//    account any other interfaces a ResponseWriter object may implement.
+//  - "decorator" or "middle": this is the user specified object that decorates an inner
```

Comment:
You generally seem to be using the term "decorator" to mean any object that adds behavior to another object.  Since it is a generic thing that happens in two ways here (one in the client, one in this package, it is confusing to _also_ use the generic term to refer to just one of the two ways it appears here.  I suggest calling the client-supplied decorator "middle" or "client-supplied decorator" (and if the latter, the contrasting object should be named in a clearly contrasting way: "implemented decorator" or "outer decorator" or something).

Reply:
I revised the comments and renamed `Decorator` to `UserProvidedDecorator`